From 0f9dda8048aae8562f289bcd1d3641e23bc7fb3f Mon Sep 17 00:00:00 2001
From: Muhammad Aksar <muhammad.aksar@intel.com>
Date: Thu, 19 Oct 2017 13:00:10 +0530
Subject: [PATCH] Storage: Avoid Crash during formatting Sdcard as Adopted
 storage

This Crash is due to returning null value while checking
for the Sdcard's Vol ID Info during making the external
storage as internal(Adopted Storage).
Avoided this java crash by implementing a small delay
before reading the Volume ID by using PublishProgress API
in Async class.

Change-Id: I8727bbd0e2cfa9bafa6d983ddec905c5e393f336
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-50574
Signed-off-by: Muhammad Aksar <muhammad.aksar@intel.com>
---
 .../android/settings/deviceinfo/StorageWizardFormatProgress.java  | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/com/android/settings/deviceinfo/StorageWizardFormatProgress.java b/src/com/android/settings/deviceinfo/StorageWizardFormatProgress.java
index 59a1866..d74edee 100644
--- a/src/com/android/settings/deviceinfo/StorageWizardFormatProgress.java
+++ b/src/com/android/settings/deviceinfo/StorageWizardFormatProgress.java
@@ -88,6 +88,7 @@ public class StorageWizardFormatProgress extends StorageWizardBase {
 
         private volatile int mProgress = 20;
 
+        private volatile long mInternalBench;
         private volatile long mPrivateBench;
 
         @Override
@@ -99,9 +100,11 @@ public class StorageWizardFormatProgress extends StorageWizardBase {
                     storage.partitionPrivate(activity.mDisk.getId());
                     publishProgress(40);
 
+                    mInternalBench = storage.benchmark(null);
+                    publishProgress(60);
+
                     final VolumeInfo privateVol = activity.findFirstVolume(VolumeInfo.TYPE_PRIVATE);
                     mPrivateBench = storage.benchmark(privateVol.getId());
-                    mPrivateBench /= 1000000;
 
                     // If we just adopted the device that had been providing
                     // physical storage, then automatically move storage to the
@@ -180,8 +183,9 @@ public class StorageWizardFormatProgress extends StorageWizardBase {
                 // benchmark, and it should be redone any time the benchmark
                 // changes.
 
+                final float pct = (float) mInternalBench / (float) mPrivateBench;
                 Log.d(TAG, "New volume took " + mPrivateBench + "ms to run benchmark");
-                if (mPrivateBench > 2000) {
+                if (Float.isNaN(pct) || pct < 0.25) {
                     final SlowWarningFragment dialog = new SlowWarningFragment();
                     dialog.showAllowingStateLoss(activity.getFragmentManager(), TAG_SLOW_WARNING);
                 } else {
-- 
1.9.1

